File: CaCompteKit/Sources/Domain/Engine/MatchEvent.swift
import Foundation

public enum MatchEvent: Sendable, Codable, Equatable {
  case matchCreated(
    gameID: String, rulesVersion: Int, variants: VariantSelection, participants: [Participant])
  case roundCommitted(RoundDraft)
  case roundAmended(index: Int, draft: RoundDraft)
  case roundRemoved(index: Int)
  case matchAbandoned(at: Date)
  /// Doc 05 « Jeu libre » et tout jeu à `manualStop` (Scrabble, Qwirkle…) : `endCheck` ne
  /// peut pas détecter cette fin tout seul, elle n'est déclenchée que par une action
  /// explicite du joueur — contrairement à `matchAbandoned`, elle produit un statut `.ended`
  /// normal (classement final écrit, comptée dans les statistiques de profil).
  case matchEndedManually
  case noteAdded(roundIndex: Int, text: String)
}

/// Un événement horodaté logiquement (horloge de Lamport) — l'unité du journal rejouable.
public struct StampedEvent: Sendable, Codable, Identifiable, Equatable {
  public let id: UUID
  public let lamport: UInt64
  public let deviceID: String
  /// Affichage uniquement, jamais pour trier : seul `(lamport, deviceID)` fait foi.
  public let occurredAt: Date
  public let event: MatchEvent

  public init(
    id: UUID = UUID(), lamport: UInt64, deviceID: String, occurredAt: Date, event: MatchEvent
  ) {
    self.id = id
    self.lamport = lamport
    self.deviceID = deviceID
    self.occurredAt = occurredAt
    self.event = event
  }
}

File: android/domain/src/main/kotlin/com/quimene/domain/engine/MatchEvent.kt
package com.quimene.domain.engine

import com.quimene.domain.model.InstantSerializer
import com.quimene.domain.model.Participant
import com.quimene.domain.model.RoundDraft
import com.quimene.domain.model.UUIDSerializer
import com.quimene.domain.model.VariantSelection
import kotlinx.serialization.Serializable
import java.time.Instant
import java.util.UUID

/**
 * Miroir de `MatchEvent.swift` — **7 cas**, pas 6 comme l'énumère par erreur le texte de
 * docs/11-portage-android.md (étape B) : le code Swift source fait foi. `Codable` côté Swift ;
 * sérialisé ici via [MatchEventSerializer], qui reproduit exactement la forme du `Codable`
 * synthétisé par Swift (`{"<cas>": {…}}`, `_0` pour un paramètre non nommé) — vérifié bit-à-bit
 * contre les golden files du protocole applicatif (dossier `spec/wire`, doc 09, étape F), pas le
 * polymorphisme à discriminant par défaut de kotlinx.serialization (`{"type": "roundCommitted",
 * ...}`) qu'aurait produit un simple `@Serializable` sur cette interface.
 */
@Serializable(with = MatchEventSerializer::class)
sealed interface MatchEvent {
    @Serializable
    data class MatchCreated(
        val gameID: String,
        val rulesVersion: Int,
        val variants: VariantSelection,
        val participants: List<Participant>,
    ) : MatchEvent

    @Serializable
    data class RoundCommitted(
        val draft: RoundDraft,
    ) : MatchEvent

    @Serializable
    data class RoundAmended(
        val index: Int,
        val draft: RoundDraft,
    ) : MatchEvent

    @Serializable
    data class RoundRemoved(
        val index: Int,
    ) : MatchEvent

    @Serializable
    data class MatchAbandoned(
        @Serializable(with = InstantSerializer::class) val at: Instant,
    ) : MatchEvent

    @Serializable
    data object MatchEndedManually : MatchEvent

    @Serializable
    data class NoteAdded(
        val roundIndex: Int,
        val text: String,
    ) : MatchEvent
}

/** Miroir de `StampedEvent.swift`. `lamport: UInt64` côté Swift → `ULong` Kotlin (support natif
 * kotlinx.serialization pour les types non signés). `occurredAt` est **documentaire, jamais
 * utilisé pour l'ordre** — seul `(lamport, deviceID)` ordonne le rejeu (`MatchEngine.replay`). */
@Serializable
data class StampedEvent(
    @Serializable(with = UUIDSerializer::class)
    val id: UUID = UUID.randomUUID(),
    val lamport: ULong,
    val deviceID: String,
    @Serializable(with = InstantSerializer::class)
    val occurredAt: Instant,
    val event: MatchEvent,
)
